sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -847,7 +847,69 @@ protected BigtableSource withEstimatedSizeBytes(Long estimatedSizeBytes) {
           Math.max(sizeEstimate / maximumNumberOfSplits, desiredBundleSizeBytes);
 
       // Delegate to testable helper.
-      return splitBasedOnSamples(desiredBundleSizeBytes, getSampleRowKeys(options));
+      List<BigtableSource> splits =
+          splitBasedOnSamples(desiredBundleSizeBytes, getSampleRowKeys(options));
+      return reduceSplits(splits, options);
+    }
+
+    private List<BigtableSource>
+        reduceSplits(List<BigtableSource> splits, PipelineOptions options) throws IOException {
+      final long maximumCountOfSplits = 15_360L;
+      int numberToCombine =
+          (int) ((splits.size() + maximumCountOfSplits - 1) / maximumCountOfSplits);
+      if (splits.size() < maximumCountOfSplits || numberToCombine < 2) {
+        return splits;
+      }
+      ImmutableList.Builder<BigtableSource> reducedSplits = ImmutableList.builder();
+      ImmutableList.Builder<ByteKeyRange> currentSplits = ImmutableList.builder();
+      int counter = 0;
+      long size = 0;
+      for (BigtableSource source : splits) {
+        size += source.getEstimatedSizeBytes(options);
+        counter++;
+        currentSplits.addAll(source.getRanges());
+        if (counter == numberToCombine) {
+          List<ByteKeyRange> rangesToCombine = currentSplits.build();
+          if (!checkRangeAdjacency(rangesToCombine)) {
```

Comment:
`checkRangeAdjancecy()` should only be applied two two ranges.  If they are adjacent, add the new Range to `rangesToCombine`; otherwise, combine what ever is in `rangesToCombine` add it to reducedSplits, and start with a new list.

Reply:
so it does not matter if we have splits of different sizes?

Reply:
@sduskis please review made suggested changes, the concern I had is that in case of non-adjacent ranges, we might have splits with sources of different sizes.